tests/testsuite/cfg.rs

```diff
@@ -490,3 +490,243 @@ fn cfg_looks_at_rustflags_for_target() {
         execs().with_status(0),
     );
 }
+
+#[test]
+fn target_cfg_bin() {
+    let p = project("target_cfg_bin")
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "bar"
+            version = "0.1.0"
+            authors = []
+
+            [[target.'cfg(target_pointer_width = "1")'.bin]]
+            name = "never"
+            path = "src/never.rs"
+
+            [[target.'cfg(not(target_pointer_width = "1"))'.bin]]
+            name = "always"
+            path = "src/always.rs"
+        "#,
+        )
+        .file("src/main.rs", &main_file("\"main\"", &[]))
+        .file("src/never.rs", &main_file("\"never\"", &[]))
+        .file("src/always.rs", &main_file("\"always\"", &[]))
+        .build();
+
+    assert_that(
+        p.cargo("build").arg("-v"),
+        execs().with_status(0)
+            .with_stderr_contains("\
+                [RUNNING] `rustc --crate-name always src[/]always.rs [..]")
+            .with_stderr_does_not_contain("\
+                [RUNNING] `rustc --crate-name never src[/]never.rs [..]"),
+    );
+}
+
+#[test]
+fn target_cfg_lib() {
+    let p = project("target_cfg_lib")
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "bar"
+            version = "0.1.0"
+            authors = []
+
+            [target.'cfg(target_pointer_width = "1")'.lib]
+            name = "never"
+            path = "src/never.rs"
+
+            [target.'cfg(not(target_pointer_width = "1"))'.lib]
+            name = "always"
+            path = "src/always.rs"
```

Comment:
I feel uneasy about the fact that you specify different names depending on the platform.... Should we make this case an error perhaps?

Reply:
As mentioned below, if they both activate, then it will be an error. No merging is done.